text/src/autogluon/text/automm/data/process_text.py

```diff
@@ -0,0 +1,262 @@
+
+        if default_segment_num <= 0:
+            default_segment_num = 1
+
+        if text_segment_num < default_segment_num:
+            warnings.warn(
+                f"provided text_segment_num: {text_segment_num} "
+                f"is smaller than {checkpoint_name}'s default: {default_segment_num}"
+            )
+        self.text_segment_num = min(text_segment_num, default_segment_num)
+        assert self.text_segment_num >= 1
+        print(f"text segment num: {self.text_segment_num}")
+
+        self.stochastic_chunk = stochastic_chunk
+
+    def collate_fn(self) -> dict:
+        fn = {}
+        fn.update({f"{self.prefix}_{TEXT_TOKEN_IDS}": Pad(pad_val=self.tokenizer.pad_token_id)})
+        fn.update({f"{self.prefix}_{TEXT_VALID_LENGTH}": Stack()})
+        fn.update({f"{self.prefix}_{TEXT_SEGMENT_IDS}": Pad(pad_val=self.tokenizer.pad_token_id)})
+        return fn
+
+    def build_one_token_sequence(
+            self,
+            text_tokens: List[NDArray[(Any,), np.int32]],
+    ) -> dict:
+
+        if self.insert_sep:
+            max_length = self.max_len - (len(text_tokens) + 1)
+        else:
+            max_length = self.max_len - 2
+        trimmed_lengths = self.get_trimmed_lengths(
+            [len(txt_token) for txt_token in text_tokens],
+            max_length,
+            do_merge=True,
+        )
+        seg = 0
+        token_ids = [self.cls_token_id]
+        segment_ids = [seg]
+        for txt_token, trim_length in zip(text_tokens, trimmed_lengths):
+
+            if self.stochastic_chunk:
+                start_ptr = np.random.randint(0, len(txt_token) - trim_length + 1)
+            else:
+                start_ptr = 0
+            token_ids.extend(txt_token[start_ptr:(start_ptr + trim_length)].tolist())
+            segment_ids.extend([seg] * trim_length)
+            if self.insert_sep:
+                token_ids.append(self.sep_token_id)
+                segment_ids.append(seg)
+            seg = (seg + 1) % self.text_segment_num
+
+        if token_ids[-1] != self.sep_token_id:
+            token_ids.append(self.sep_token_id)
+            segment_ids.append(seg)
+
+        return {
+            f"{self.prefix}_{TEXT_TOKEN_IDS}": np.array(token_ids, dtype=np.int32),
+            f"{self.prefix}_{TEXT_VALID_LENGTH}": len(token_ids),
+            f"{self.prefix}_{TEXT_SEGMENT_IDS}": np.array(segment_ids, dtype=np.int32)
+        }
+
+    def build_one_token_sequence_from_text(
+            self,
+            text: List[str],
+    ) -> dict:
+        # tokenize text
+        tokens = []
+        warnings.filterwarnings(
+            "ignore",
+            "Token indices sequence length is longer than.*result in indexing errors"
+        )
+        for col_text in text:
+            col_tokens = self.tokenizer.encode(
+                col_text,
+                add_special_tokens=False,
+                truncation=False,
+            )
+            tokens.append(np.array(col_tokens, dtype=np.int32))
```

Comment:
Is there any good reason why we work with numpy instead of working directly with torch tensors? Seems inefficient and memory consuming to me.

Reply:
The main reason is to use NDArray in the type hint to control array shape. We don't find memory issue using the current implementation. We can leave this to the future discussion whether it's necessary to use torch.tensor.